linkerd/proxy/api-resolve/src/resolve.rs

```diff
@@ -88,69 +85,71 @@ where
+where
+    S: GrpcService<BoxBody> + Clone + Send + 'static,
+    S::Error: Into<Box<dyn Error + Send + Sync>> + Send,
+    S::ResponseBody: Send,
+    <S::ResponseBody as Body>::Data: Send,
+    <S::ResponseBody as HttpBody>::Error: Into<Box<dyn Error + Send + Sync + 'static>> + Send,
+    S::Future: Send,
+{
+    try_stream! {
+        let rsp = client.get(grpc::Request::new(req)).await?;
+        trace!(metadata = ?rsp.metadata());
+        let mut stream = rsp.into_inner();
+        while let Some(update) = stream.next().await {
+            match update?.update {
+                Some(api::update::Update::Add(api::WeightedAddrSet {
+                    addrs,
+                    metric_labels,
+                })) => {
+                    let addr_metas = addrs
+                        .into_iter()
+                        .filter_map(|addr| pb::to_addr_meta(addr, &metric_labels))
+                        .collect::<Vec<_>>();
+                    if !addr_metas.is_empty() {
+                        debug!(endpoints = %addr_metas.len(), "Add");
+                        yield Update::Add(addr_metas);
                     }
+                }
 
-                    Some(api::update::Update::Remove(api::AddrSet { addrs })) => {
-                        let sock_addrs = addrs
-                            .into_iter()
-                            .filter_map(pb::to_sock_addr)
-                            .collect::<Vec<_>>();
-                        if !sock_addrs.is_empty() {
-                            debug!(endpoints = %sock_addrs.len(), "Remove");
-                            return Poll::Ready(Some(Ok(Update::Remove(sock_addrs))));
-                        }
+                Some(api::update::Update::Remove(api::AddrSet { addrs })) => {
+                    let sock_addrs = addrs
+                        .into_iter()
+                        .filter_map(pb::to_sock_addr)
+                        .collect::<Vec<_>>();
+                    if !sock_addrs.is_empty() {
+                        debug!(endpoints = %sock_addrs.len(), "Remove");
+                        yield Update::Remove(sock_addrs);
                     }
+                }
 
-                    Some(api::update::Update::NoEndpoints(api::NoEndpoints { exists })) => {
-                        info!("No endpoints");
-                        let update = if exists {
-                            Update::Empty
-                        } else {
-                            Update::DoesNotExist
-                        };
-                        return Poll::Ready(Some(Ok(update.into())));
-                    }
+                Some(api::update::Update::NoEndpoints(api::NoEndpoints { exists })) => {
+                    info!("No endpoints");
+                    let update = if exists {
+                        Update::Empty
+                    } else {
+                        Update::DoesNotExist
+                    };
+                    yield update.into();
```

Comment:
nit, take it or leave it: can this just be
```suggestion
                    if exists {
                        yield Update::Empty.into();
                    }
                    yield Update::DoesNotExist.into();
```
not a blocker.

Reply:
I might be unclear on the semantics of _yield_, but this at least seems like it would always yield DNE after yielding Empty?

In either case, the original shows the intent much more clearly imo.